driver/src/orderbook/streamed/balance.rs

```diff
@@ -63,29 +89,91 @@ impl Balance {
     }
 
     pub fn get_balance(&self, batch_id: BatchId) -> Result<U256> {
-        let mut balance = self.balance_with_deposit(batch_id)?;
+        let mut balance = self.balance_with_deposit_and_proceeds(batch_id)?;
         if self.withdraw.batch_id < batch_id {
             // Saturating because withdraw requests can be for amounts larger than balance.
             balance = balance.saturating_sub(self.withdraw.amount);
         }
         Ok(balance)
     }
 
-    fn balance_with_deposit(&self, current_batch_id: BatchId) -> Result<U256> {
+    pub fn sell(&mut self, amount: u128, batch_id: BatchId) -> Result<()> {
+        self.sell_buy_internal(amount, batch_id, TradeType::Sell)
+    }
+
+    pub fn buy(&mut self, amount: u128, batch_id: BatchId) -> Result<()> {
+        self.sell_buy_internal(amount, batch_id, TradeType::Buy)
+    }
+
+    pub fn revert_sell(&mut self, amount: u128, batch_id: BatchId) -> Result<()> {
+        self.revert_sell_buy_internal(amount, batch_id, TradeType::Sell)
+    }
+
+    pub fn revert_buy(&mut self, amount: u128, batch_id: BatchId) -> Result<()> {
+        self.revert_sell_buy_internal(amount, batch_id, TradeType::Buy)
+    }
+
+    fn sell_buy_internal(
```

Comment:
nit: I would have preferred if this was `buy_sell_internal` since nobody, anywhere, has ever said "sell-buy".